File: Sorting/quickSort.c
//Implementation of Quick Sort

#include<stdio.h>
void input(int A[],int n)
{
    int i;
    printf("Enter %d number of elements: ",n);
    for(i=0;i<n;i++)
        scanf("%d",&A[i]);
}
void display(int A[],int n)
{
    int i;
    for(i=0;i<n;i++)
        printf("%d ",A[i]);
}
void swap(int *ptr1,int *ptr2)
{
    int temp=*ptr1;
    *ptr1=*ptr2;
    *ptr2=temp;
}
int partition(int A[],int p,int r)
{
    int x=A[r];
    int i=p-1;
    for(int j=p;j<r;j++)
    {
        if(A[j]<=x)
        {
            i++;
            swap(&A[i],&A[j]);
        }
    }
    swap(&A[i+1],&A[r]);
    return i+1;
}
void QuickSort(int A[],int p,int r)
{
    if(p<r)
    {
        int q=partition(A,p,r);
        QuickSort(A,p,q-1);
        QuickSort(A,q+1,r);
    }
}
void main()
{
    int A[20],n;
    printf("\nEnter the size of the array:");
    scanf("%d",&n);
    input(A,n);
    printf("\nArray before sorting:");
    display(A,n);
    QuickSort(A,0,n-1);
    printf("\nArray after sorting:");
    display(A,n);
}

/* Output:

Enter	the size of the array:10
Enter 10 number of elements: 5 7 20 3 11 6 15 4 10 12

Array	before sorting:5 7 20 3 11 6 15 4 10 12
Array	after sorting:3 4 5 6 7 10 11 12 15 20
*/

File: Sorting/QuickSort.java
import java.util.Arrays;
import java.util.Scanner;

public class QuickSort {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter array size: ");


        //taking input from the user
        int n=sc.nextInt();
        int[] array=new int[n];
        System.out.println("Enter array elements space separated");
        for(int i=0;i<n;i++){
            array[i]=sc.nextInt();

        }

        //call to the quicksort function
        quickSort(array,0,array.length);

        System.out.println(Arrays.toString(array));


    }


    public static int Partition(int[] A,int l,int h){
        int pivot=A[l];
        int i=l;
        int j=h;
        while(i<j){
            do {
                i++;
            }while(i<j && A[i]<=pivot);

            do{
                j--;
            }while(A[j]>pivot);
            if(i<j)
                swap(A,i,j);

        }
        swap(A,l,j);

        return j;
    }

    public static void swap(int[] A,int a,int b){

        // function to swap to numbers a and b in array.
        int temp=A[a];
        A[a]=A[b];
        A[b]=temp;

    }

    public static void quickSort(int[] array,int l, int h){
        if(l<h){
            int j=Partition(array,l,h);

            quickSort(array,l,j);
            quickSort(array,j+1,h);
        }

    }
}
